scripts/igmp_query_injector.py

```diff
@@ -0,0 +1,459 @@
+            ret.extend(self.parse_mac_address(mac))
+
+        if self.vlanid == -1:
+            type_field = [0x08, 0x00]
+        else:
+            type_field = [0x81, 0x00]
+
+        ret.extend(type_field)
+        return ret
+
+    def create_vlan_layer(self):
+        if self.vlanid == -1:
+            return []
+        else:
+            return [0x20 | (self.vlanid >> 8), self.vlanid & 0xff, 0x08, 0x00]
+
+    def create_ip_layer(self):
+        """
+        IP Type: IPv4 (0x0800)
+        Version: 4
+        Total length: 32
+        TTL: 1
+        Source IP: 0.0.0.0
+        Destination IP: 224.0.0.1 (e0:00:00:01)
+        """
+        return [
+            0x46, 0x00, 0x00, 0x20, 0x00, 0x01,
+            0x00, 0x00, 0x01, 0x02, 0x44, 0xd6,
+            0x00, 0x00, 0x00, 0x00, 0xe0, 0x00,
+            0x00, 0x01, 0x94, 0x04, 0x00, 0x00,
+        ]
+
+    def create_igmp_layer(self):
+        """
+        IGMP Version: 2
+        IGMP Type: IGMP Query (0x11)
+        Max Resp Time: self.max_resp_time
+        Multicast Address: 0.0.0.0
+        """
+        msg_type = 0x11
+        # MAX_RESP_TIME filed unit is 100ms
+        max_resp_time = self.max_resp_time / 100
+        group_address = [0x00, 0x00, 0x00, 0x00]
+        chksum = self._calc_igmp_chksum(msg_type, max_resp_time, group_address)
+        ret = [msg_type, max_resp_time] + chksum + group_address
+        return ret
+
+    def _calc_igmp_chksum(self, msg_type, max_resp_time, group_address):
+        msg = [msg_type, max_resp_time, 0x00, 0x00] + group_address
+        chksum = socket.htons(self._checksum(msg))
+        return [(chksum & 0xff00) >> 8, chksum & 0xff]
+
+    def _checksum(self, msg):
+        def carry_around_add(a, b):
+            c = a + b
+            return (c & 0xffff) + (c >> 16)
+
+        s = 0
+        for i in range(0, len(msg), 2):
+            w = msg[i] + (msg[i+1] << 8)
+            s = carry_around_add(s, w)
+        return ~s & 0xffff
+
+    def generate(self):
+        ether_layer = self.create_ether_layer()
+        vlan_layer = self.create_vlan_layer()
+        ip_layer = self.create_ip_layer()
+        igmp_layer = self.create_igmp_layer()
+        packet = ether_layer + vlan_layer + ip_layer + igmp_layer
+        return "".join(map(chr, packet))
+
+
+class IGMPQueryInjector(object):
+    __metaclass__ = ABCMeta
+
+    def __init__(self, max_resp_time):
+        self.max_resp_time = max_resp_time
+
+    def inject_to_vif(self, vif, mac):
+        log.info('Inject IGMP query to vif:%s, mac:%s' % (vif, mac))
+        packet = IGMPQueryGenerator(mac, -1, self.max_resp_time).generate()
+        self.inject_query_packet(vif, packet)
+
+    def inject_to_vifs(self, vifs):
+        for vif in vifs:
+            _vif = Vif(vif)
+            self.inject_to_vif(_vif.vif_name, _vif.mac)
+
+    def inject_query_packet(self, interface, packet):
+        s = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, 0)
```

Comment:
It would be better to open the socket once and send all the query packets, rather than destroying and recreating it each time.

Reply:
The IGMP query send to different interfaces, so we have to recreate it every time.